pulsar-broker/src/test/java/org/apache/pulsar/broker/delayed/bucket/BucketDelayedDeliveryTrackerThreadSafetyTest.java

```diff
@@ -0,0 +1,412 @@
+import org.testng.annotations.Test;
+
+/**
+ * Thread safety tests for BucketDelayedDeliveryTracker.
+ * These tests verify that the hybrid approach with StampedLock and concurrent data structures
+ * correctly handles concurrent access patterns without deadlocks, race conditions, or data corruption.
+ */
+public class BucketDelayedDeliveryTrackerThreadSafetyTest {
+
+    private BucketDelayedDeliveryTracker tracker;
+    private AbstractPersistentDispatcherMultipleConsumers dispatcher;
+    private ManagedCursor cursor;
+    private Timer timer;
+    private BucketSnapshotStorage storage;
+    private ExecutorService executorService;
+
+    @BeforeMethod
+    public void setUp() throws Exception {
+        dispatcher = mock(AbstractPersistentDispatcherMultipleConsumers.class);
+        cursor = mock(ManagedCursor.class);
+        timer = mock(Timer.class);
+        storage = mock(BucketSnapshotStorage.class);
+
+        when(dispatcher.getName()).thenReturn("persistent://public/default/test-topic / test-cursor");
+        when(dispatcher.getCursor()).thenReturn(cursor);
+        when(cursor.getName()).thenReturn("test-cursor");  // Provide a valid cursor name
+        when(cursor.getCursorProperties()).thenReturn(java.util.Collections.emptyMap());
+
+        // Mock cursor property operations for bucket key storage
+        when(cursor.putCursorProperty(any(), any()))
+                .thenReturn(CompletableFuture.completedFuture(null));
+        when(cursor.removeCursorProperty(any()))
+                .thenReturn(CompletableFuture.completedFuture(null));
+
+        // Mock storage operations to avoid NullPointerException
+        when(storage.createBucketSnapshot(any(), any(), any(), any(), any()))
+                .thenReturn(CompletableFuture.completedFuture(1L));
+        when(storage.getBucketSnapshotMetadata(anyLong()))
+                .thenReturn(CompletableFuture.completedFuture(null));
+        when(storage.getBucketSnapshotSegment(anyLong(), anyLong(), anyLong()))
+                .thenReturn(CompletableFuture.completedFuture(java.util.Collections.emptyList()));
+        when(storage.getBucketSnapshotLength(anyLong()))
+                .thenReturn(CompletableFuture.completedFuture(0L));
+        when(storage.deleteBucketSnapshot(anyLong()))
+                .thenReturn(CompletableFuture.completedFuture(null));
+
+        tracker = new BucketDelayedDeliveryTracker(
+            dispatcher, timer, 1000, Clock.systemUTC(), true, storage,
+            100, 1000, 100, 10  // Restore original minIndexCountPerBucket for proper testing
+        );
+
+        executorService = Executors.newFixedThreadPool(32);
+    }
+
+    @AfterMethod
+    public void tearDown() throws Exception {
+        if (tracker != null) {
+            tracker.close();
+        }
+        if (executorService != null) {
+            executorService.shutdown();
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
```

Comment:
Guava `com.google.common.util.concurrent.MoreExecutors`'s `shutdownAndAwaitTermination` could be used for shutdown here instead. It contains logic to first call shutdown and then shutdownNow. It's also good to use `assertTrue` to make sure that the executor really got shutdown too. We don't use this pattern in our tests currently, but we could start doing that to avoid possible thread leaks in tests.